web/source/kmwnative.js

```diff
@@ -94,22 +94,30 @@
   * 
   * @param  {Object}  Ln  keyboard stub object    
   */    
-  keymanweb.keyboardUnavailable = function(Ln)
-  {
-    return window.setTimeout(function() {
-        util.wait(false);
-        var Ps=keymanweb._KeyboardStubs[Ln],kbdName=Ps['KN'],lgName=Ps['KL'];
-        kbdName=kbdName.replace(/\s*keyboard\s*/i,'');
+  keymanweb.keyboardUnavailable = function(Ln) {
+    keymanweb.loadFailureHandler = function(altString) {
```

Comment:
It sounds like that would mean significant changes to the functionality of KR, sounding an absolute death-knell for raw inclusion of a keyboard's scripting file without stub info.  I mean, we require stub info right now, but we haven't always required it.

It would be an absolute requirement for a keyboard to be loaded via stub and to be loaded only when actually set active.

That being said, yes, it is quite ugly that loading is resolved in KR.  I think what you mention in #2 for an object would be extremely useful and could form the basis of an object-valued 'kbdstub.loadingState' in line with your point #3.

That could facilitate asynchronous keyboard loading, which is subject to one lone issue:  the util.alert and until.wait functions aren't modal and could rapidly change if multiple loading errors arise, causing loss of feedback and confusion.